Sources/DeepLearning/Trainer.swift

```diff
@@ -0,0 +1,47 @@
+// Copyright 2018 The TensorFlow Authors. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+/// SupervisedTrainer implements a standard training loop for a model.
+public struct SupervisedTrainer<Model, Opt: Optimizer, Loss: Differentiable>
+    where Opt.Model == Model,
+          Loss: FloatingPoint,
+          Loss == Loss.CotangentVector {
+    // TODO: Convert to protocol to take advantage of argument labels to disambiguate.
+    typealias LossFn = @differentiable (Model.Output, Model.Output) -> Loss
+
+    var model: Model
+    var optimizer: Opt
```

Comment:
If `Optimizer` is guaranteed to be a class then this can be `let`.

Reply:
Marking as resolved, as the code has been completely restructured.